src/http/input/body/N3PatchBodyParser.ts

```diff
@@ -0,0 +1,136 @@
+import type { DefaultGraph, NamedNode, Quad, Quad_Subject } from '@rdfjs/types';
+import { DataFactory, Parser, Store } from 'n3';
+import type { Term } from 'rdf-js';
+import { getBlankNodes, getTerms, getVariables, mapTerms } from 'rdf-terms';
+import { TEXT_N3 } from '../../../util/ContentTypes';
+import { BadRequestHttpError } from '../../../util/errors/BadRequestHttpError';
+import { createErrorMessage } from '../../../util/errors/ErrorUtil';
+import { UnprocessableEntityHttpError } from '../../../util/errors/UnprocessableEntityHttpError';
+import { UnsupportedMediaTypeHttpError } from '../../../util/errors/UnsupportedMediaTypeHttpError';
+import { guardedStreamFrom, readableToString } from '../../../util/StreamUtil';
+import { RDF, SOLID } from '../../../util/Vocabularies';
+import type { N3Patch } from '../../representation/N3Patch';
+import type { BodyParserArgs } from './BodyParser';
+import { BodyParser } from './BodyParser';
+
+/**
+ * Parses an N3 Patch document and makes sure it conforms to the specification requirements.
+ * Requirements can be found at Solid, §5.3.1: https://solid.github.io/specification/protocol#n3-patch
+ */
+export class N3PatchBodyParser extends BodyParser {
+  public async canHandle({ metadata }: BodyParserArgs): Promise<void> {
+    if (metadata.contentType !== TEXT_N3) {
+      throw new UnsupportedMediaTypeHttpError('This parser only supports N3 Patch data.');
+    }
+  }
+
+  public async handle({ request, metadata }: BodyParserArgs): Promise<N3Patch> {
+    const n3 = await readableToString(request);
+    const parser = new Parser({ format: TEXT_N3, baseIRI: metadata.identifier.value });
+    let store: Store;
+    try {
+      store = new Store(parser.parse(n3));
+    } catch (error: unknown) {
+      throw new BadRequestHttpError(`Invalid N3: ${createErrorMessage(error)}`);
+    }
+    const def = DataFactory.defaultGraph();
+
+    // Solid, §5.3.1: "A patch resource MUST contain a triple ?patch rdf:type solid:InsertDeletePatch."
+    // "The patch document MUST contain exactly one patch resource,
+    // identified by one or more of the triple patterns described above, which all share the same ?patch subject."
+    const subjects = store.getSubjects(RDF.terms.type, SOLID.terms.InsertDeletePatch, def);
+    if (subjects.length !== 1) {
+      throw new UnprocessableEntityHttpError('An N3 Patch document requires exactly 1 solid:InsertDeletePatch entry.');
+    }
+    const subject = subjects[0];
```

Comment:
Suggestion: perhaps this line and below should be a separate method `applyPatch`, such that we can easily extend to multi-patch documents.